flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java

```diff
@@ -0,0 +1,145 @@
+
+import java.util.Map;
+
+/**
+ * Rowtime descriptor for describing an event time attribute in the schema.
+ */
+@PublicEvolving
+public class Rowtime implements Descriptor {
+
+	private final DescriptorProperties internalProperties = new DescriptorProperties(true);
+
+	public static final String ROWTIME = "rowtime";
+	public static final String ROWTIME_TIMESTAMPS_TYPE = "rowtime.timestamps.type";
+	public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD = "from-field";
+	public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE = "from-source";
+	public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM = "custom";
+	public static final String ROWTIME_TIMESTAMPS_FROM = "rowtime.timestamps.from";
+	public static final String ROWTIME_TIMESTAMPS_CLASS = "rowtime.timestamps.class";
+	public static final String ROWTIME_TIMESTAMPS_SERIALIZED = "rowtime.timestamps.serialized";
+
+	public static final String ROWTIME_WATERMARKS_TYPE = "rowtime.watermarks.type";
+	public static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING = "periodic-ascending";
+	public static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED = "periodic-bounded";
+	public static final String ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE = "from-source";
+	public static final String ROWTIME_WATERMARKS_TYPE_VALUE_CUSTOM = "custom";
+	public static final String ROWTIME_WATERMARKS_CLASS = "rowtime.watermarks.class";
+	public static final String ROWTIME_WATERMARKS_SERIALIZED = "rowtime.watermarks.serialized";
+	public static final String ROWTIME_WATERMARKS_DELAY = "rowtime.watermarks.delay";
+
+	/**
+	 * Sets a built-in timestamp extractor that converts an existing {@link Long} or
+	 * {@link Types#SQL_TIMESTAMP} field into the rowtime attribute.
+	 *
+	 * @param fieldName The field to convert into a rowtime attribute.
+	 */
+	public Rowtime timestampsFromField(String fieldName) {
+		internalProperties.putString(ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD);
+		internalProperties.putString(ROWTIME_TIMESTAMPS_FROM, fieldName);
+		return this;
+	}
+
+	/**
+	 * Sets a built-in timestamp extractor that converts the assigned timestamps from
+	 * a DataStream API record into the rowtime attribute and thus preserves the assigned
+	 * timestamps from the source.
+	 *
+	 * <p>Note: This extractor only works in streaming environments.
+	 */
+	public Rowtime timestampsFromSource() {
+		internalProperties.putString(ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE);
+		return this;
+	}
+
+	/**
+	 * Sets a custom timestamp extractor to be used for the rowtime attribute.
+	 *
+	 * @param extractor The {@link TimestampExtractor} to extract the rowtime attribute
+	 *                  from the physical type.
+	 */
+	public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
+		internalProperties.putString(ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM);
```

Comment:
I think this changes the property format that we use underneath. How about we add method `TimestampExtractor#toProperties()`? The default implementation could be like this:

```
	protected Map<String, String> toProperties() {
		Map<String, String> properties = new HashMap<>();
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM);
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> extractor.getClass.getName);
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> EncodingUtils.encodeObjectToString(extractor));
		return properties;
	}
```

In the classes that used custom format (e.g. ExisitngField) we can still use more specific properties.

Reply:
Good idea!

Reply:
Sorry, I forgot this. I will try my best not to happen this again.